modules/chargebee-integration/app/webhook.go

```diff
@@ -32,46 +33,90 @@ func WebhookHandler(baseLogger *zap.SugaredLogger, mapping []map[string]string)
 		}
 
 		logger := baseLogger.With("event id", event["id"])
-		if event["event_type"] != SubscriptionCreated {
+		responseCode := http.StatusOK
+		switch event["event_type"] {
+		case SubscriptionCreated:
+			responseCode = handleSubscriptionCreated(logger, event, mapping)
+		case SubscriptionCancelled:
+			go handleSubscriptionCancelled(logger, event, mapping)
+		default:
 			logger.Errorf("event type is unsupported: %s\n", event["event_type"])
-			w.WriteHeader(http.StatusOK)
-			return
 		}
+		w.WriteHeader(responseCode)
+	}
+}
 
-		subscription, err := GetSubscription(event)
-		if err != nil {
-			logger.Error("error retrieving subscription", err)
-			w.WriteHeader(http.StatusBadRequest)
-			return
-		}
+func handleSubscriptionCreated(logger *zap.SugaredLogger, event map[string]interface{}, mapping []map[string]string) int {
+	subscription, err := GetSubscription(event)
+	if err != nil {
+		logger.Error("error retrieving subscription", err)
+		return http.StatusBadRequest
+	}
+
+	logger = logger.With("subscription id", subscription.Id)
+	logger.Infof("subscription status: %s", subscription.Status)
 
-		logger = logger.With("subscription id", subscription.Id)
-		logger.Infof("subscription status: %s", subscription.Status)
+	svc := createServiceItem()
+	svc.Subscription = subscription.Id
 
-		svc := createServiceItem()
-		svc.Subscription = subscription.Id
+	err = ApplyMapping(logger, subscription, mapping, svc)
+	if err != nil {
+		logger.Error("error applying mapping", err)
+		return http.StatusBadRequest
+	}
 
-		err = ApplyMapping(logger, subscription, mapping, svc)
+	err = createService(logger, svc)
+	if err != nil && checkAlreadyExists(err) {
+		logger.Error("service already exists", err)
+		// expected behavior due to prevent retries https://www.chargebee.com/docs/2.0/events_and_webhooks.html
+		return http.StatusOK
+	} else if err != nil {
+		logger.Error("create operation error", err)
+		return http.StatusServiceUnavailable
+	}
+
+	return http.StatusOK
+}
+
+func handleSubscriptionCancelled(logger *zap.SugaredLogger, event map[string]interface{}, mapping []map[string]string) {
+	subscriptionId, err := GetSubscriptionId(event)
+	if err != nil {
+		logger.Error("Error extracting subscription from payload", err)
+		return
+	}
+	logger = logger.With("subscription id", subscriptionId)
+	// Check if service exists
+	service, err := getServiceBySubscriptionId(logger, subscriptionId)
```

Comment:
We need to retry when something wrong happens during this step: there might be an issue getting service even when it exists.

Reply:
Fixed